### PR TITLE
Improving Test coverage

### PR DESCRIPTION
**Pull request by apurvabanka**
| Name                                     | Stmts | Miss | Branch | BrPart | Cover |
|------------------------------------------|-------|------|--------|--------|-------|
| fastkml/__init__.py                      | 27    | 0    | 0      | 0      | 100%  |
| fastkml/about.py                         | 1     | 0    | 0      | 0      | 100%  |
| fastkml/atom.py                          | 65    | 0    | 0      | 0      | 100%  |
| fastkml/base.py                          | 75    | 0    | 14     | 0      | 100%  |
| fastkml/config.py                        | 23    | 0    | 4      | 0      | 100%  |
| fastkml/containers.py                    | 60    | 0    | 4      | 0      | 100%  |
| fastkml/data.py                          | 110   | 4    | 2      | 0      | 96%   |
| fastkml/enums.py                         | 81    | 0    | 30     | 0      | 100%  |
| fastkml/exceptions.py                    | 5     | 0    | 0      | 0      | 100%  |
| fastkml/features.py                      | 141   | 3    | 4      | 0      | 98%   |
| fastkml/geometry.py                      | 268   | 1    | 68     | 0      | 99%   |
| fastkml/gx.py                            | 139   | 2    | 52     | 0      | 99%   |
| fastkml/helpers.py                       | 184   | 0    | 72     | 0      | 100%  |
| fastkml/kml.py                           | 69    | 0    | 10     | 1      | 99%   |
| fastkml/kml_base.py                      | 22    | 0    | 0      | 0      | 100%  |
| fastkml/links.py                         | 46    | 0    | 0      | 0      | 100%  |
| fastkml/mixins.py                        | 18    | 0    | 0      | 0      | 100%  |
| fastkml/overlays.py                      | 158   | 6    | 0      | 0      | 96%   |
| fastkml/registry.py                      | 45    | 0    | 8      | 2      | 96%   |
| fastkml/styles.py                        | 196   | 4    | 4      | 0      | 98%   |
| fastkml/times.py                         | 103   | 3    | 30     | 6      | 93%   |
| fastkml/types.py                         | 13    | 0    | 0      | 0      | 100%  |
| fastkml/views.py                         | 125   | 5    | 0      | 0      | 96%   |

@cleder Coverage Report

Related Issue #352

## Summary by Sourcery

Enhance test coverage by adding new test cases across multiple modules, focusing on error handling and utility functions.

Tests:
- Add new test cases to improve coverage for various modules, including containers, kml, geometries, gx, styles, overlays, and views.
- Introduce tests for handling GeometryError exceptions in multiple geometry-related modules.
- Add tests for helper functions in a new helper_test.py file to ensure proper functionality of utility methods.
- Add tests for geometry functions in a new functions_test.py file to validate error handling and coordinate processing.

## Summary by CodeRabbit

- **New Features**
	- Enhanced test coverage for the `PhotoOverlay` class, ensuring correct initialization and attribute validation.
	- Introduced new tests for KML container functionalities, including feature appending and style URL retrieval.
	- Added tests for error handling in boundary geometry classes.
	- Established a comprehensive testing framework for geometry-related functions, addressing invalid geometries.
	- Introduced a new test for the `Track` class to validate the `etree_element` method.
	- Improved assertions in the `MultiTrack` and `Region` tests for better validation of object states.

- **Bug Fixes**
	- Corrected assertions in various tests to ensure accurate validation of attributes.

- **Documentation**
	- Updated comments for clarity in test methods related to the `Region` class.

### CODE REVIEW

File: tests/gx_test.py [hunk 1]
```diff
@@ -330,6 +337,8 @@ def test_from_multilinestring(self) -> None:
 
         mt = MultiTrack(geometry=lines, ns="")
 
+        assert mt.__bool__() is True
```
Comment (cleder) [inline tests/gx_test.py:340]: I think
```suggestion
        assert bool(mt)
```
is easier to read

File: tests/overlays_test.py [hunk 2]
```diff
@@ -249,6 +249,8 @@ def test_create_photo_overlay_with_all_optional_parameters(self) -> None:
         assert photo_overlay.description == "This is a photo overlay"
         assert photo_overlay.shape == enums.Shape.rectangle
         assert photo_overlay.rotation == 0
+        assert photo_overlay.view_volume.__bool__() is True
+        assert photo_overlay.image_pyramid.__bool__() is True
```
Comment (cleder) [inline tests/overlays_test.py:253]:
```suggestion
        assert bool(photo_overlay.view_volume)
        assert bool(photo_overlay.image_pyramid)
```

File: tests/views_test.py [hunk 3]
```diff
@@ -209,6 +209,7 @@ def test_region_with_all_optional_parameters(self) -> None:
         assert region.lod.min_fade_extent == 0
         assert region.lod.max_fade_extent == 512
         assert region
+        assert region.lod.__bool__() is True
```
Comment (cleder) [inline tests/views_test.py:212]: I think `assert region`  does the same as 
```suggestion
        assert bool(region)
```

File: tests/containers_test.py [hunk 4]
```diff
@@ -65,6 +66,37 @@ def test_document_boolean_visibility_invalid(self) -> None:
...
+
+    def test_container_feature_append(self) -> None:
+        container = containers._Container(
+            ns="ns",
+            id="id",
+            target_id="target_id",
+            name="name"
+        )
+        feature = features._Feature(name="new_feature")
+        assert container.append(feature) is None
```
Comment (cleder) [inline tests/containers_test.py:87]: I don't think this line is needed, as `.append` always returns `None`

**Review by cleder — CHANGES_REQUESTED** [review]
Hey, nice work :+1: 
I left some comments and :+1: or :-1: on the AI generated reviews.

**Comment by apurvabanka** [comment]
> Hey, nice work 👍 I left some comments and 👍 or 👎 on the AI generated reviews.

@cleder Thanks for the review. Just to clarify, does  👍 mean we are good with the changes or should I go ahead and make the change suggested by the AI bot?

**Comment by cleder** [comment]
Sorry @apurvabanka a :-1: on a bot review means not needed, a :+1:  means, it is a good suggestion.

File: tests/containers_test.py (hunk 4 above)
Comment (apurvabanka) [reply tests/containers_test.py]: changed

File: tests/containers_test.py [hunk 5]
```diff
@@ -65,6 +66,37 @@ def test_document_boolean_visibility_invalid(self) -> None:
...
+        with pytest.raises(ValueError):
+            container.append(container)
```
[the remark this answers is not in the document]
Comment (apurvabanka) [reply tests/containers_test.py]: implemented

File: tests/geometries/boundaries_test.py [hunk 6]
```diff
@@ -41,6 +43,16 @@
...
+    def test_outer_boundry_geometry_error(self) -> None:
+        """Test GeometryError."""
+        p = geo.Point(1, 2)
+        coords = ((1, 2), (2, 0), (0, 0), (1, 2))
+
+        with pytest.raises(GeometryError):
+            OuterBoundaryIs(
+                kml_geometry=LinearRing(kml_coordinates=Coordinates(coords=coords)),
+                geometry=p
+            )
```
[the remark this answers is not in the document]
Comment (apurvabanka) [reply tests/geometries/boundaries_test.py]: implemented

**Comment by apurvabanka** [comment]
@cleder Are there any pending items for this PR?

**Comment by cleder** [comment]
you did not see that mypy failed?
```
tests/kml_test.py:180: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
tests/kml_test.py:200: error: Argument 1 to "append" of "KML" has incompatible type "KML"; expected "Folder | Document | Placemark | GroundOverlay | PhotoOverlay"  [arg-type]
tests/helper_test.py:24: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
tests/helper_test.py:26: error: "node_text" does not return a value (it only ever returns None)  [func-returns-value]
tests/helper_test.py:27: error: Argument "obj" to "node_text" has incompatible type "None"; expected "_XMLObject"  [arg-type]
tests/helper_test.py:28: error: Argument "element" to "node_text" has incompatible type "None"; expected "Element"  [arg-type]
tests/helper_test.py:32: error: Argument "verbosity" to "node_text" has incompatible type "int"; expected "Verbosity"  [arg-type]
tests/helper_test.py:38: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
tests/helper_test.py:40: error: "float_attribute" does not return a value (it only ever returns None)  [func-returns-value]
tests/helper_test.py:41: error: Argument "obj" to "float_attribute" has incompatible type "None"; expected "_XMLObject"  [arg-type]
tests/helper_test.py:42: error: Argument "element" to "float_attribute" has incompatible type "str"; expected "Element"  [arg-type]
tests/helper_test.py:46: error: Argument "verbosity" to "float_attribute" has incompatible type "int"; expected "Verbosity"  [arg-type]
tests/helper_test.py:47: error: Argument "default" to "float_attribute" has incompatible type "str"; expected "float | None"  [arg-type]
tests/helper_test.py:52: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
tests/helper_test.py:54: error: "enum_attribute" does not return a value (it only ever returns None)  [func-returns-value]
tests/helper_test.py:55: error: Argument "obj" to "enum_attribute" has incompatible type "None"; expected "_XMLObject"  [arg-type]
tests/helper_test.py:56: error: Argument "element" to "enum_attribute" has incompatible type "str"; expected "Element"  [arg-type]
tests/helper_test.py:60: error: Argument "verbosity" to "enum_attribute" has incompatible type "int"; expected "Verbosity"  [arg-type]
tests/helper_test.py:61: error: Argument "default" to "enum_attribute" has incompatible type "str"; expected "Enum | None"  [arg-type]
tests/helper_test.py:65: error: Function is missing a return type annotation  [no-untyped-def]
tests/helper_test.py:65: note: Use "-> None" if function does not return a value
tests/helper_test.py:81: error: Function is missing a return type annotation  [no-untyped-def]
tests/helper_test.py:81: note: Use "-> None" if function does not return a value
tests/helper_test.py:98: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
tests/helper_test.py:105: error: Argument "name_spaces" to "attribute_float_kwarg" has incompatible type "str"; expected "dict[str, str]"  [arg-type]
tests/helper_test.py:108: error: Argument "classes" to "attribute_float_kwarg" has incompatible type "None"; expected "tuple[type[_XMLObject] | type[Enum] | type[bool] | type[int] | type[str] | type[float], ...]"  [arg-type]
tests/helper_test.py:115: error: Incompatible types in assignment (expression has type "None", variable has type "str")  [assignment]
tests/helper_test.py:[12](https://github.com/cleder/fastkml/actions/runs/11316861155/job/31488710465#step:5:13)1: error: Argument "name_spaces" to "subelement_enum_kwarg" has incompatible type "str"; expected "dict[str, str]"  [arg-type]
tests/helper_test.py:124: error: Argument "classes" to "subelement_enum_kwarg" has incompatible type "list[type[Color]]"; expected "tuple[type[_XMLObject] | type[Enum] | type[bool] | type[int] | type[str] | type[float], ...]"  [arg-type]
tests/helper_test.py:[13](https://github.com/cleder/fastkml/actions/runs/11316861155/job/31488710465#step:5:14)5: error: Argument "name_spaces" to "attribute_enum_kwarg" has incompatible type "str"; expected "dict[str, str]"  [arg-type]
tests/helper_test.py:138: error: Argument "classes" to "attribute_enum_kwarg" has incompatible type "list[type[Color]]"; expected "tuple[type[_XMLObject] | type[Enum] | type[bool] | type[int] | type[str] | type[float], ...]"  [arg-type]
tests/containers_test.py:96: error: Argument "style_url" to "Document" has incompatible type "str"; expected "StyleUrl | None"  [arg-type]
tests/geometries/multigeometry_test.py:306: error: Argument "geometry" to "MultiGeometry" has incompatible type "Point"; expected "MultiPoint | MultiLineString | MultiPolygon | GeometryCollection | None"  [arg-type]
tests/geometries/multigeometry_test.py:306: error: Argument "kml_geometries" to "MultiGeometry" has incompatible type "Coordinates"; expected "Iterable[Point | LineString | Polygon | LinearRing | MultiGeometry] | None"  [arg-type]
tests/geometries/linestring_test.py:47: error: Argument "geometry" to "LineString" has incompatible type "Point"; expected "LineString | None"  [arg-type]
tests/geometries/functions_test.py:13: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
tests/geometries/functions_test.py:[17](https://github.com/cleder/fastkml/actions/runs/11316861155/job/31488710465#step:5:18): error: Argument "error" to "handle_invalid_geometry_error" has incompatible type "type[ValueError]"; expected "Exception"  [arg-type]
tests/geometries/functions_test.py:23: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
tests/geometries/functions_test.py:25: error: "handle_invalid_geometry_error" does not return a value (it only ever returns None)  [func-returns-value]
tests/geometries/functions_test.py:26: error: Argument "error" to "handle_invalid_geometry_error" has incompatible type "type[ValueError]"; expected "Exception"  [arg-type]
tests/geometries/functions_test.py:47: error: Argument "node_name" to "coordinates_subelement" has incompatible type "None"; expected "str"  [arg-type]
tests/geometries/functions_test.py:63: error: "coordinates_subelement" does not return a value (it only ever returns None)  [func-returns-value]
tests/geometries/functions_test.py:64: error: Argument "obj" to "coordinates_subelement" has incompatible type "None"; expected "_XMLObject"  [arg-type]
tests/geometries/functions_test.py:66: error: Argument "node_name" to "coordinates_subelement" has incompatible type "None"; expected "str"  [arg-type]
tests/geometries/boundaries_test.py:78: error: Function is missing a type annotation  [no-untyped-def]
tests/geometries/boundaries_test.py:88: error: Function is missing a return type annotation  [no-untyped-def]
tests/geometries/boundaries_test.py:88: note: Use "-> None" if function does not return a value
tests/geometries/boundaries_test.py:92: error: Function is missing a return type annotation  [no-untyped-def]
tests/geometries/boundaries_test.py:92: note: Use "-> None" if function does not return a value
Found 45 errors in 7 files (checked 56 source files)
```

**Comment by cleder** [comment]
I also had to improve some test, so they were asserting something non-trivial, and removed tests that were not of value

**Comment by apurvabanka** [comment]
@cleder For my PR I just ran the Test Coverage for the pytest. I might have missed some setup steps. How do you run mypy?

**Comment by cleder** [comment]
See [contributing.rst](https://github.com/cleder/fastkml/blob/develop/docs/contributing.rst#setting-up-your-environment)

```
pre-commit run --all
```

**Pull request merged**